substrate/network-libp2p/src/service.rs

```diff
@@ -0,0 +1,942 @@
+				//		 beforehand?
+				let peer_id = match shared.network_state.accept_custom_proto(node_id.clone(), protocol_id, custom_proto_out.protocol_version, endpoint, custom_proto_out.outgoing) {
+					Ok(peer_id) => peer_id,
+					Err(err) => return future::Either::A(future::err(err.into())),
+				};
+
+				debug!(target: "sub-libp2p", "Successfully connected to {:?} (peer id \
+						{}) with protocol {:?} version {}",
+					node_id, peer_id, protocol_id, custom_proto_out.protocol_version);
+				handler.connected(&NetworkContextImpl {
+					inner: shared.clone(),
+					protocol: protocol_id,
+					current_peer: Some(peer_id),
+				}, &peer_id);
+
+				struct KadDisconnectGuard(Arc<Shared>, PeerId, PeerstorePeerId, Arc<NetworkProtocolHandler + Send + Sync>, ProtocolId);
+				impl Drop for KadDisconnectGuard {
+					fn drop(&mut self) {
+						debug!(target: "sub-libp2p", "Node {:?} with peer ID {} \
+							through protocol {:?} disconnected", self.2, self.1,
+							self.4);
+						self.3.disconnected(&NetworkContextImpl {
+							inner: self.0.clone(),
+							protocol: self.4,
+							current_peer: Some(self.1),
+						}, &self.1);
+
+						// When any custom protocol drops, we drop the peer entirely.
+						// TODO: is this correct?
+						self.0.network_state.disconnect_peer(self.1);
+					}
+				}
+				let dc_guard = KadDisconnectGuard(shared.clone(), peer_id, node_id.clone(), handler.clone(), protocol_id);
+
+				future::Either::B(custom_proto_out
+					.incoming
+					.for_each(move |(packet_id, data)| {
+						shared.kad_system.update_kbuckets(node_id.clone());
+						handler.read(&NetworkContextImpl {
+							inner: shared.clone(),
+							protocol: protocol_id,
+							current_peer: Some(peer_id.clone()),
+						}, &peer_id, packet_id, &data);
+						Ok(())
+					})
+					.then(move |val| {
+						// Makes sure that `dc_guard` is kept alive until here.
+						drop(dc_guard);
+						val
+					}))
+			})) as Box<_>
+		},
+	}
+}
+
+// Builds the multiaddress corresponding to the address we need to listen to according to the
+// config.
+// TODO: put the `Multiaddr` directly in the `NetworkConfiguration`
+fn config_to_listen_addr(config: &NetworkConfiguration) -> Multiaddr {
+	if let Some(addr) = config.listen_address {
+		let ip = match addr.ip() {
+			IpAddr::V4(addr) => AddrComponent::IP4(addr),
+			IpAddr::V6(addr) => AddrComponent::IP6(addr),
+		};
+		iter::once(ip).chain(iter::once(AddrComponent::TCP(addr.port()))).collect()
+	} else {
+		let host = AddrComponent::IP4(Ipv4Addr::new(0, 0, 0, 0));
+		let port = AddrComponent::TCP(0);
+		iter::once(host).chain(iter::once(port)).collect()
+	}
+}
+
+// Randomly discovers peers to connect to.
+// This works by running a round at a regular interval, and skipping if we reached `min_peers`.
+// When we are over `min_peers`, we stop trying to dial nodes and only accept incoming
+// connections.
+fn start_kademlia_discovery<T, To, St, C>(shared: Arc<Shared>, transport: T,
```

Comment:
this function goes 11(!) indent levels deep in multiple places. definitely could be broken up into isolatable components.